tests/unit/test_http_client.py

```diff
@@ -0,0 +1,363 @@
+
+
+def test_http_client_raises_json(tmpdir, caplog):
+    """Client initialization raises when keyfile not valid json."""
+    caplog.set_level(logging.DEBUG)
+
+    tmp_keyfile = tmpdir.mkdir('keys').join('broken_keyfile.json')
+    tmp_keyfile.write('broken json')
+
+    with pytest.raises(exceptions.GCPGordonJanitorError) as e:
+        http_client.AIOGoogleHTTPClient(keyfile=tmp_keyfile)
+
+    e.match(f'Keyfile {tmp_keyfile} is not valid JSON.')
+    assert 1 == len(caplog.records)
+
+
+def test_http_client_raises_not_found(tmpdir, caplog):
+    """Client initialization raises when keyfile not found."""
+    caplog.set_level(logging.DEBUG)
+
+    tmp_keydir = tmpdir.mkdir('keys')
+    no_keyfile = os.path.join(tmp_keydir, 'not-existent.json')
+
+    with pytest.raises(exceptions.GCPGordonJanitorError) as e:
+        http_client.AIOGoogleHTTPClient(keyfile=no_keyfile)
+
+    e.match(f'Keyfile {no_keyfile} was not found.')
+    assert 1 == len(caplog.records)
+
+
+#####
+# Tests & fixtures for access token handling
+#####
+@pytest.fixture
+def mock_parse_expiry(mocker, monkeypatch):
+    mock = mocker.MagicMock(oauth_client, autospec=True)
+    mock._parse_expiry.return_value = datetime.datetime(2018, 1, 1, 12, 0, 0)
+    monkeypatch.setattr('gordon_janitor_gcp.http_client._client', mock)
+    return mock
+
+
+@pytest.fixture
+def client(fake_keyfile, mock_credentials):
+    client = http_client.AIOGoogleHTTPClient(keyfile=fake_keyfile)
+    yield client
+    # test teardown
+    client._session.close()
+
+
+@pytest.mark.asyncio
+async def test_refresh_token(client, fake_keyfile_data, mock_parse_expiry,
+                             caplog):
+    """Successfully refresh access token."""
+    caplog.set_level(logging.DEBUG)
+
+    url = fake_keyfile_data['token_uri']
+    token = 'c0ffe3'
+    payload = {
+        'access_token': token,
+        'expires_in': 3600,  # seconds = 1hr
+    }
+    with aioresponses() as mocked:
+        mocked.post(url, status=200, payload=payload)
+        await client.refresh_token()
+    assert token == client.token
+    assert 2 == len(caplog.records)
+
+
+args = 'status,payload,exc,err_msg'
+params = [
+    [504, None, exceptions.GCPHTTPError, 'Issue connecting to example.com'],
+    [200, {}, exceptions.GCPAuthError, 'No access token in response.'],
+]
+
+
+@pytest.mark.parametrize(args, params)
+@pytest.mark.asyncio
+async def test_refresh_token_raises(status, payload, exc, err_msg, client,
+                                    fake_keyfile_data, caplog):
+    """Response errors from attempting to refresh token."""
+    caplog.set_level(logging.DEBUG)
+
+    url = fake_keyfile_data['token_uri']
+
+    with aioresponses() as mocked:
+        mocked.post(url, status=status, payload=payload)
+        with pytest.raises(exc) as e:
+            await client.refresh_token()
+
+        e.match(err_msg)
+
+    assert 3 == len(caplog.records)
+
+
+# pytest prevents monkeypatching datetime directly
```

Comment:
I haven't used it much, but I've seen [freezegun](https://github.com/spulec/freezegun) used in similar tests to make a date easier to set up and cleanup. I don't think it's an issue here & now, but the test changes and leaves `datetime.datetime` set to the mock.

This is thinking aloud, I'm fine with either approach here btw.

Reply:
I actually tried using freezegun, but I couldn't figure out how to patch the client I was testing (I assume this is an issue on the user side 😬 ), and also it only supports `datetime.datetime.now()` and now `datetime.datetime.utcnow()`.